server/stats.go

```diff
@@ -0,0 +1,228 @@
+package main
+
+import (
+	"encoding/json"
+	goexpvar "expvar"
+	"math/rand"
+	"net/http"
+	"os"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/mattermost/mattermost-plugin-jira/server/expvar"
+	"github.com/pkg/errors"
+)
+
+const statsKeyExpiration = 30 * 24 * time.Hour // 30 days
+const statsAutosaveInterval = 10 * time.Minute
+const statsAutosaveMaxDither = 60 // seconds
+
+var initStatsOnce sync.Once
+
+func (p *Plugin) initStats() {
+	conf := p.getConfig()
+	if conf.DisableStats || conf.stats != nil {
+		return
+	}
+
+	data, appErr := p.API.KVGet(statsKeyName())
+	if appErr != nil {
+		return
+	}
+	stats := expvar.NewStats(data)
+
+	p.updateConfig(func(c *config) {
+		initStatsOnce.Do(func() {
+			if !c.DisableStats {
+				c.stats = stats
+			}
+		})
+	})
+
+	initUserCounter(p.currentInstanceStore, p.userStore)
+	initUptime()
+
+	p.startAutosaveStats()
+}
+
+func httpAPIStats(p *Plugin, w http.ResponseWriter, r *http.Request) (int, error) {
+	if r.Method != http.MethodGet {
+		return http.StatusMethodNotAllowed,
+			errors.New("method " + r.Method + " is not allowed, must be GET")
+	}
+
+	conf := p.getConfig()
+
+	//TODO protect from unauthorized access?
+	// status, err := verifyWebhookRequestSecret(conf, r)
+	// if err != nil {
+	// 	return status, err
+	// }
+
+	if conf.stats == nil {
+		return http.StatusNotFound, errors.New("No stats available")
+	}
+
+	out := "{"
+	first := true
+	conf.stats.Do(func(name string, e *expvar.Endpoint) {
+		if first {
+			first = false
+		} else {
+			out += ","
+		}
+		out += `"` + name + `":` + e.String()
```

Comment:
This will break of the strings contain invalid json. It would be better to collect them into a map then use json marshal.

Reply:
This is intentional. `e` implements String() as specified in https://golang.org/pkg/expvar/#Var, so it returns a valid JSON (summary) value, as a `string`. Short of either using a different interface, or re-parsing the JSON, this implementation seems simplest (and valid).

Reply:
You could use `Value` rather than `String`. But this works since the string is valid JSON 👍